internal/services/bootstrap.go

```diff
@@ -120,16 +118,51 @@ func (s *Service) FillStakerAddr(ctx context.Context, numWorkers int, dryRun boo
 // It extracts events from each block and forwards them to the event processor.
 // Returns an error if it fails to get block results or process events.
 func (s *Service) processBlocksSequentially(ctx context.Context) error {
+	ctx, cancel := context.WithCancelCause(ctx)
```

Comment:
This is first time I'm using new context API `WithCancelCause(...)`, previously there was only `WithCancel(...)`. Looks quite suitable